draft-aegis-mimi-arch-latest.md

```diff
@@ -43,28 +43,166 @@ TODO Abstract
+' +-------+     +----------------------+       |                '     |
+' |  Bob  | --- | Application Server 2 |-------+                '     |
+' +-------+     +----------------------+                        '     |
+'                                                               '     |
++ - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - +     |
+                                                                      |
++ - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - +     |
+'                       Domain B                                '     |
+'                                                               '     |
+' +-------+     +----------------------+     +----------------+ '     |
+' | Carol | --- | Application Server 3 | --- | MIMI Gateway B |-'-----+ 
+' +-------+     +----------------------+     +----------------+ '
+' +-------+     +----------------------+       |                '
+' | Dave  | --- | Application Server 4 |-------+                '
+' +-------+     +----------------------+                        '
+'                                                               '
++ - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - +
+~~~
 
-Define the interface.
+# Gateway Overview
+
+The MIMI Gateway is designed to be a hybrid of AS and DS functionality has
+defined by `MLS Architecture`. It has an internal interface for
+applications within its domain to federate with each other, as well as an
+external interface that allows for cross-domain federation. A gateway acts as a
+permissioned database of the following information:
+
+* Unique identities that can be grouped by tags.
+* Queryable queues of MLS Key Packages indexed by identity, protocol version, cipher suite, custom labels.
+* MLS Group proposals, commits, welcome messages and ratchet trees indexed by epoch.
+
+## Identity Providers
+
+An identity provider helps ensure a consistent AS behavior between federating
+MLS applications. Internally an identity provider MUST implement the following
+set of functionality. 
+
+* Declare support for a specific set of MLS credential types.
+* Validate a credential given a public key that is stored along side that
+  credential in an MLS Leaf and a timestamp that can be used to determine
+  validity of credentials that may expire (Ex. X509 Certificate credentials).
+* Uniquely identify a Leaf based on its identity to ensure that the same
+  identity is not added to a MLS group multiple times.
+* Uniquely identity the entity that controls a specific leaf.
+* Determine if an identity is controlled by the same entity as another identity.
+
+An example interface of an identity provider is as follows:
+
+~~~
+struct {
+    HPKEPublicKey public_key;
+    Credential credential;
+} Identity;
+
+fn supported_credentials() -> [CredentialType];
+fn validate(Identity identity, u64 time) -> bool;
+fn leaf_identifier(Identity identity) -> Vec<u8>;
+fn entity_identifier(Identity identity) -> Vec<u8>;

```

Comment:
I really hate bringing this up... The RFC ended up saying
> the LeafNode in the path field of the external commit MUST meet the same criteria as would the LeafNode in an Update for the removed leaf

So it looks like we only need one thing here. Maybe the leaf cred doesn't matter at all; it just holds an ephemeral signature key?

Reply:
Lets chat about this, I'm not sure what you are getting at. The primary reason for having that identifier is to make sure the same client doesn't get added twice to a convo.